contracts/0.4.24/Lido.sol

```diff
@@ -7,255 +7,326 @@ pragma solidity 0.4.24;
+        uint256 timeElapsed
+    ) external;
+}
 
+interface ILidoExecutionLayerRewardsVault {
+    function withdrawRewards(uint256 _maxAmount) external returns (uint256 amount);
+}
+
+interface IWithdrawalVault {
+    function withdrawWithdrawals(uint256 _amount) external;
+}
 
-interface IERC721 {
-    /// @notice Transfer ownership of an NFT
-    /// @param _from The current owner of the NFT
-    /// @param _to The new owner
-    /// @param _tokenId The NFT to transfer
-    function transferFrom(address _from, address _to, uint256 _tokenId) external payable;
+interface IStakingRouter {
+    function deposit(
+        uint256 maxDepositsCount,
+        uint256 stakingModuleId,
+        bytes depositCalldata
+    ) external payable returns (uint256);
+    function getStakingRewardsDistribution()
+        external
+        view
+        returns (
+            address[] memory recipients,
+            uint256[] memory stakingModuleIds,
+            uint96[] memory stakingModuleFees,
+            uint96 totalFee,
+            uint256 precisionPoints
+        );
+    function getWithdrawalCredentials() external view returns (bytes32);
+    function reportRewardsMinted(uint256[] _stakingModuleIds, uint256[] _totalShares) external;
 }
 
+interface IWithdrawalQueue {
+    function finalizationBatch(uint256 _lastRequestIdToFinalize, uint256 _shareRate)
+        external
+        view
+        returns (uint128 eth, uint128 shares);
+    function finalize(uint256 _lastIdToFinalize) external payable;
+    function isPaused() external view returns (bool);
+    function unfinalizedStETH() external view returns (uint256);
+    function isBunkerModeActive() external view returns (bool);
+}
 
 /**
 * @title Liquid staking pool implementation
 *
-* Lido is an Ethereum 2.0 liquid staking protocol solving the problem of frozen staked Ethers
-* until transfers become available in Ethereum 2.0.
-* Whitepaper: https://lido.fi/static/Lido:Ethereum-Liquid-Staking.pdf
-*
-* NOTE: the code below assumes moderate amount of node operators, e.g. up to 200.
+* Lido is an Ethereum liquid staking protocol solving the problem of frozen staked ether on Consensus Layer
+* being unavailable for transfers and DeFi on Execution Layer.
 *
 * Since balances of all token holders change when the amount of total pooled Ether
 * changes, this token cannot fully implement ERC20 standard: it only emits `Transfer`
 * events upon explicit transfer between holders. In contrast, when Lido oracle reports
 * rewards, no Transfer events are generated: doing so would require emitting an event
 * for each token holder and thus running an unbounded loop.
-*
-* At the moment withdrawals are not possible in the beacon chain and there's no workaround.
-* Pool will be upgraded to an actual implementation when withdrawals are enabled
-* (Phase 1.5 or 2 of Eth2 launch, likely late 2022 or 2023).
 */
-contract Lido is ILido, StETH, AragonApp {
+contract Lido is StETHPermit, AragonApp, Versioned {
```

Comment:
`Versioned, StETHPermit, AragonApp`

Reply:
check storage layout and write some tests

Reply:
fixed the order and added the doc, tests should be included into the test plan